ofrak_core/ofrak/model/component_model.py

```diff
@@ -17,6 +19,62 @@ class ComponentConfig:
     """
 
 
+@dataclass
+class ComponentExternalTool:
+    """
+    An external tool or utility (like `zip` or `squashfs`) a component depends on
+    """
+
+    tool: str
+    install_hints: str = None  # e.g. version
+    install_check_arg: str = "--help"
+
+    async def run_tool(self, *args: str) -> str:
+        """
+        Run an external CLI tool this component depends on, as a subprocess.
+
+        :param args: arguments to give the tool
+
+        :return: standard output if the subprocess completes successfully.
+
+        :raises ComponentMissingDependencyError: If the OS indicates it cannot find the tool.
+        :raises ComponentSubprocessError: If the subprocess fails (any return other than 0).
+        """
+        try:
+            proc = await asyncio.create_subprocess_exec(
+                self.tool, *args, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
+            )
+        except FileNotFoundError:
+            raise ComponentMissingDependencyError(self.tool, self.install_hints)
+
+        proc_stdout, proc_stderr = await proc.communicate()
+        if proc.returncode == 0:
+            return proc_stdout.decode("ascii")
+        else:
+            raise ComponentSubprocessError(
+                self.tool,
+                args,
+                proc.returncode,
+                proc_stdout.decode("ascii"),
+                proc_stderr.decode("ascii"),
+            )
+
+    async def is_tool_installed(self) -> bool:
+        try:
+            proc = await asyncio.create_subprocess_exec(
+                self.tool,
+                self.install_check_arg,
+            )
+        except FileNotFoundError:
+            return False
+
+        retcode = await proc.wait()
+        if retcode == 0:
+            return True
+        else:
+            return False
+
```

Comment:
I'd consider also adding a `install_tool` command.

A couple other things to consider, is that we have different types of tools

1. Command line tools (`pigz`, `tar`, for example)
2. Python tools with custom installation steps (`binwalk`; binary ninja)
3. Tools that are not invoked directly at the command line: ghidra, for example.

Ideally this class would handle all of these cases

Reply:
For this reason I want to stay away from a general "install tool" command. For tools that can be easily installed via a package manager like `brew` or `apt`, I've added a standard way to define that.

Reply:
We decided not to install tools within OFRAK as part of this PR.